src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_GetLatestDiagnosticsForSpan.cs

```diff
@@ -107,18 +109,137 @@ private class LatestDiagnosticsForSpanGetter
+            {
+                // unfortunately, we need to special case compiler diagnostic analyzer so that 
+                // we can do span based analysis even though we implemented it as semantic model analysis
+                if (stateSet.Analyzer == _compilerAnalyzer)
+                {
+                    return await TryGetSyntaxAndSemanticCompilerDiagnostics(stateSet, textVersion, syntaxVersion, semanticVersion).ConfigureAwait(false);
+                }
+
+                var containsFullResult = await TryGetDocumentDiagnosticsAsync(
+                                            stateSet, StateType.Syntax, (t, d) => t.Equals(textVersion) && d.Equals(syntaxVersion), GetSyntaxDiagnosticsAsync).ConfigureAwait(false);
+
+                containsFullResult &= await TryGetDocumentDiagnosticsAsync(
+                    stateSet, StateType.Document, (t, d) => t.Equals(textVersion) && d.Equals(semanticVersion), GetSemanticDiagnosticsAsync).ConfigureAwait(false);
+
+                return containsFullResult;
+            }
+
+            private async Task<bool> TryGetSyntaxAndSemanticCompilerDiagnostics(StateSet stateSet, VersionStamp textVersion, VersionStamp syntaxVersion, VersionStamp semanticVersion)
+            {
+                // First, get syntax errors
+                var containsFullResult = await TryGetDocumentDiagnosticsAsync(
+                                            stateSet, StateType.Syntax, true,
+                                            (t, d) => t.Equals(textVersion) && d.Equals(syntaxVersion),
+                                            async (_1, _2) =>
+                                            {
+                                                var root = await _document.GetSyntaxRootAsync(_cancellationToken).ConfigureAwait(false);
+                                                var diagnostics = root.GetDiagnostics();
+
+                                                return GetDiagnosticData(_document, root.SyntaxTree, _range, diagnostics);
+                                            }).ConfigureAwait(false);
+
+                // second get semantic errors
+                containsFullResult &= await TryGetDocumentDiagnosticsAsync(
+                    stateSet, StateType.Document, true,
+                    (t, d) => t.Equals(textVersion) && d.Equals(semanticVersion),
+                    async (_1, _2) =>
+                    {
+                        var model = await _document.GetSemanticModelAsync(_cancellationToken).ConfigureAwait(false);
+                        VerifyDiagnostics(model);
+
+                        var root = await _document.GetSyntaxRootAsync(_cancellationToken).ConfigureAwait(false);
+                        var adjustedSpan = AdjustSpan(_document, root, _range);
+                        var diagnostics = model.GetDeclarationDiagnostics(adjustedSpan, _cancellationToken).Concat(model.GetMethodBodyDiagnostics(adjustedSpan, _cancellationToken));
```

Comment:
syntax doesnt support span based get diagnostics. considering how parser works, it probably doesnt make sense to support such diagnostic since it doesnt have a way to do so anyway.